g3doc/Tutorials/Toy/Ch-2.md

```diff
@@ -23,139 +23,320 @@ addresses this issue by being designed for extensibility. As such, there are
+  /// This interface is used by the builder to allow for easily generating
+  /// instances of this operation:
+  ///   mlir::OpBuilder::create<ConstantOp>(...)
+  /// This method populates the given `state` that MLIR uses to create
+  /// operations. This state is a collection of all of the discrete elements
+  /// that an operation may contain.
+  /// Build a constant with the given return type and `value` attribute.
+  static void build(mlir::Builder *builder, mlir::OperationState *state,
+                    mlir::Type result, mlir::DenseElementsAttr value);
+  /// Build a constant and reuse the type from the given 'value'.
+  static void build(mlir::Builder *builder, mlir::OperationState *state,
+                    mlir::DenseElementsAttr value);
+  /// Build a constant by broadcasting the given 'value'.
+  static void build(mlir::Builder *builder, mlir::OperationState *state,
+                    double value);
+
+  /// Return the value of the constant by fetching it from the attribute.
+  mlir::DenseElementsAttr getValue();
+};
 ```
 
-This approach is used in `Ch2/mlir/MLIRGen.cpp` to implement a naive MLIR
-generation through a simple depth-first search traversal of the Toy AST. Here is
-how we create a `toy.transpose` operation:
+and we register this operation in the `ToyDialect` constructor:
 
 ```c++
-mlir::Operation *createTransposeOp(OpBuilder &builder,
-                                   mlir::Value *input_tensor) {
-  // Fill the `OperationState` with the required fields.
-  mlir::OperationState result(location, "toy.transpose");
-  result.addOperands(input_tensor);
-
-  // We use the MLIR tensor type for 'toy' types.
-  auto type = builder.getTensorType({2, 2}, builder.getF64Type());
-  result.addTypes(type);
-
-  // Create the transpose operation.
-  Operation *newTransposeOp = builder->createOperation(result);
-  return newTransposeOp;
+ToyDialect::ToyDialect(mlir::MLIRContext *ctx)
+    : mlir::Dialect(getDialectNamespace(), ctx) {
+  addOperations<ConstantOp>();
+}
+```
+
+### Op vs Operation: Using MLIR Operations
+
+Now that we have defined an operation, we will want to access and transform it.
+In MLIR, there are two main classes related to operations: `Operation` and `Op`.
+Operation is the actual opaque instance of the operation, and represents the
+general API into an operation instance. An `Op` is the base class of a derived
+operation, like `ConstantOp`, and acts as smart pointer wrapper around a
+`Operation*`. This means that when we define our Toy operations, we are actually
+providing a clean interface for building and interfacing with the `Operation`
+class; this is why our `ConstantOp` defines no class fields. Therefore, we
+always pass these classes around by-value, instead of by reference or pointer
+(passing by-value is a common idiom and applies similarly to attributes, types,
+etc). We can always get an instance of our toy operation by using LLVM's casting
+infrastructure:
```

Comment:
This kind of paragraph is exactly why I prefer to keep things separated: now we're not only introducing IR concept, we also introduce a bunch of details around the software engineering structure, mixed with implementation detail (link missing for LLVM casting infrastructure by the way).